DBAL-95 Firebird Driver, Platform and Schema Manager

ancpru opened the pull request:
Hi, 

I've implemented the driver for Firebird. It's currently tested with FB 2.5.

The driver uses the ibase-api. Originally I also had a PDO-based implementation. Despite it worked quite well in a real world application, it terribly failed in the Doctrine test suite because of strange transaction behaviour of the Firebird-PDO, thus I finally removed it.

The name of the driver is ibase_firebird, the platform name is firebird.

Because of the common ancestor Firebird an Interbase, it _should_ be possible to use the implementation for Interbase, too. But this is not done nore tested, yet.

Andreas

ancpru committed:
Merge pull request #3 from doctrine/master

upd

Commit:
Firebird driver, platform and schema added

Commit:
Firebird Schema Test Class added

Commit:
Keywords added and firebird-driver registered in DriverManager

ancpru committed:
Merge pull request #5 from doctrine/master

Upd from HEAD

Commit:
Many Changes in Firebird Driver and tests

Commit:
Merge composer/DBAL-95-firebird into DBAL-95-firebird

Commit:
Directory structure of ibase-drivers changed (IbaseFirebird went into ibase/Firebird),
AbstractFbIbDriverTest modified to skip other drivers

Commit:
Enabled ibase driver to handle named parameters (simulated by SQLParserUtils)

Commit:
Removed PDOFirebird driver because of instability

Commit:
Updated some comments

Commit:
Some more Firebird tests

kimhemsoe commented on lib/Doctrine/DBAL/Driver/AbstractFbIbDriver.php:
```diff
@@ -0,0 +1,198 @@
+     * 
+     * @var type 
+     */
+    protected $driverOptions = array();
```
Attributes should only be protected if needed.

ancpru replied:
I do not understand. Should they rather be public or private?

I usually set them protected if a direct assignment would not have any effect.

kimhemsoe replied:
Private if it make sense. 
Protected should be avoided. Protected just give us another api to maintain in the future.

ancpru replied:
I see. And in this case completely removed because unused anyway.

kimhemsoe replied:
Did you want to remove this or?

kimhemsoe commented on lib/Doctrine/DBAL/Driver/AbstractFbIbDriver.php:
```diff
@@ -0,0 +1,198 @@
+        $message = 'Error ' . $exception->getErrorCode() . ': ' . $message;
+        switch ($exception->getErrorCode()) {
+            case -104: {
+                    return new \Doctrine\DBAL\Exception\SyntaxErrorException($message, $exception);
```
Maybe alias the namespace with a use statement.

kimhemsoe commented on lib/Doctrine/DBAL/Driver/AbstractFbIbDriver.php:
```diff
@@ -0,0 +1,198 @@
+        return new \Doctrine\DBAL\Exception\DriverException($message, $exception);
+
+
+        switch ($exception->getErrorCode()) {
```
This look like dead code to me.

ancpru replied:
Yep, removed.

kimhemsoe commented on lib/Doctrine/DBAL/Driver/Ibase/AbstractIbaseStatement.php:
```diff
@@ -0,0 +1,672 @@
+    /**
+     * @var resource
+     */
+    protected $_dbh;
```
No need for prefixing with underscore.

ancpru replied:
Was unused anayway because the Statement gets the handle over the Connection-instance anyway.

kimhemsoe commented on lib/Doctrine/DBAL/Driver/Ibase/AbstractIbaseStatement.php:
```diff
@@ -0,0 +1,672 @@
+        $this->connection = $connection;
+        $this->ibaseStatementRc = null;
+        $this->affectedRows = false;
+        $this->numFields = false;
```
Are these 3 lines needed ?

kimhemsoe commented on lib/Doctrine/DBAL/Driver/Ibase/AbstractIbaseStatement.php:
```diff
@@ -0,0 +1,672 @@
+     */
+    public function bindParam($column, &$variable, $type = null, $length = null)
+    {
+        if (is_array($variable))
```
When you are seeing this ? dbal connection should handle this for you.

ancpru replied:
Which part do you mean exactly?

kimhemsoe replied:
That $variable is an array. Dbal\Connection will expand it, before it reaches the drivers.

ancpru replied:
Well, actually this is just the case if the query is performed via executeQuery() or executeUpdate(), but not if somebody goes the   ->prepare() ... ->bindParam()  ... ->execute() route. It's just here to offer list-parameters in such statements, too. But I do not know if other driver offer this.

kimhemsoe replied:
They do not.

Some drivers don't even support named parameters, only positional.

ancpru replied:
The ibase-api itself does not offer named parameters, too, but I simulate them using an internal name => position map because they are much more convinient. And real pepared statements are better for the firebird server. 

But if none of the other drivers support lists it might be better to remove support for them in the firebird-driver, too, as the support of lists adds complexity and causes a special case where a naive prepare cannot be used anyway because of the parameter-dependant necessary query-rewrite.

kimhemsoe commented on lib/Doctrine/DBAL/Driver/Ibase/AbstractIbaseStatement.php:
```diff
@@ -0,0 +1,672 @@
+    protected function internalFetchBoth()
+    {
+        $tmpData = ibase_fetch_assoc($this->ibaseResultRc, IBASE_TEXT);
+        if (!$tmpData === FALSE)
```
Can be replaced wih !==

Commit:
Removed dead parts, removed support for list parameters in Firebird driver, and some more cleanups

Commit:
Fixed incorrect handling of single quotes in SQL-Strings escaped by repeated single-quote (DBAL-1205)

ancpru committed:
Merge pull request #6 from helicon-os/DBAL-1205

Fixed incorrect handling of single quotes in SQL-Strings escaped by repe...

Commit:
Fixed missing tmp_view_name in drop block statement

Commit:
Merge composer/DBAL-95-firebird into DBAL-95-firebird

Commit:
Merged with DBAL-1205 fix in order to handle execute block statement correctly

Commit:
Removed inPdoContext platform attribute because not longer needed

ancpru commented:
Hm, I do not really understand why my latest commits partly fail in the travis test with strange errors like the following with drivers I did not even touch and run w/o error in firebird and mysql in my local development environment

1) Doctrine\Tests\DBAL\Functional\BlobTest::testInsert

Doctrine\DBAL\Exception\ConnectionException: An exception occured in driver: SQLSTATE[HY000] [2002] No such file or directory

/home/travis/build/doctrine/dbal/lib/Doctrine/DBAL/Driver/AbstractMySQLDriver.php:103

/home/travis/build/doctrine/dbal/lib/Doctrine/DBAL/DBALException.php:133

/home/travis/build/doctrine/dbal/lib/Doctrine/DBAL/Driver/PDOMySql/Driver.php:47

/home/travis/build/doctrine/dbal/lib/Doctrine/DBAL/Connection.php:362

/home/travis/build/doctrine/dbal/lib/Doctrine/DBAL/Connection.php:431

/home/travis/build/doctrine/dbal/lib/Doctrine/DBAL/Connection.php:391

/home/travis/build/doctrine/dbal/lib/Doctrine/DBAL/Connection.php:328

/home/travis/build/doctrine/dbal/tests/Doctrine/Tests/TestUtil.php:86

/home/travis/build/doctrine/dbal/tests/Doctrine/Tests/TestUtil.php:52

/home/travis/build/doctrine/dbal/tests/Doctrine/Tests/TestUtil.php:43

/home/travis/build/doctrine/dbal/tests/Doctrine/Tests/DbalFunctionalTestCase.php:43

/home/travis/build/doctrine/dbal/tests/Doctrine/Tests/DBAL/Functional/BlobTest.php:15

Caused by

Doctrine\DBAL\Driver\PDOException: SQLSTATE[HY000] [2002] No such file or directory

ancpru committed:
Merge pull request #7 from doctrine/master

Pull changes from Head Doctrine Repository

wizhippo committed:
Fix internalBeginTransaction resource check and Interbase transaction wait support

Interbase only supports "wait" or "no wait". timeout not support.
Should call checkLastApiCall if NOT a resource.

wizhippo committed:
Allow infinite timeout support for interbase

ancpru committed:
Merge pull request #8 from wizhippo/patch-1

Fix internalBeginTransaction resource check and Interbase transaction…

ancpru committed:
Merge pull request #10 from doctrine/master

merge head

wizhippo committed:
Add charset support

franbenz committed:
Changed getListTablesSQL condition.
This doc http://www.firebirdfaq.org/faq174/ suggests using this new condition. Some old tables have RDB$RELATION_TYPE = NULL.

franbenz committed:
All field names where lower cased, so field names on mappings should also be lower cased.
This helps the command doctrine:mapping:import match entity fields to its own mapping field names.

ancpru committed:
Merge pull request #13 from franbenz/firebird-tables

Changed getListTablesSQL condition.

ancpru committed:
Merge pull request #14 from franbenz/mapping-field-case-sensitivity

All field names where lower cased, so field names on mappings should also be lower cased.

ancpru committed:
Merge pull request #12 from wizhippo/add-charset-support

Add charset support

ancpru committed:
Merge pull request #16 from doctrine/master

Head merged back

mathielen commented:
+1

kimhemsoe commented on lib/Doctrine/DBAL/Driver/AbstractFbIbDriver.php:
```diff
@@ -0,0 +1,154 @@
+    /**
+     * Returns the driver options
+     */
+    public function getDriverOptions()
```
Is this function needed?

kimhemsoe commented on lib/Doctrine/DBAL/Driver/AbstractFbIbDriver.php:
```diff
@@ -0,0 +1,154 @@
+     * 
+     * @param array $optionsToSet
+     */
+    public function setDriverOptions(array $optionsToSet)
```
Seems this method could be private?

kimhemsoe commented on lib/Doctrine/DBAL/Driver/AbstractFbIbDriver.php:
```diff
@@ -0,0 +1,154 @@
+     * @param type $password
+     * @param array $driverOptions
+     */
+    public function connect(array $params, $username = null, $password = null, array $driverOptions = array())
```
It seems very weird to me that it does not return a connection as defined by the interface

kimhemsoe commented on lib/Doctrine/DBAL/Driver/Ibase/AbstractIbaseConnection.php:
```diff
@@ -0,0 +1,500 @@
+ * 
+ * @author Andreas Prucha, Helicon Software Development <prucha@helicon.co.at>
+ */
+abstract class AbstractIbaseConnection implements Connection, ServerInfoAwareConnection
```
I do not see any reason why all the attributes are protected. Is it not possible to change them all to private?

kimhemsoe commented on lib/Doctrine/DBAL/Driver/Ibase/AbstractIbaseConnection.php:
```diff
@@ -0,0 +1,500 @@
+    /**
+     * @var string Charset
+     */
+    protected $charset = null;
```
Any reason why you save dbs, username, password and charset on the connection object?

kimhemsoe commented on lib/Doctrine/DBAL/Driver/Ibase/AbstractIbaseConnection.php:
```diff
@@ -0,0 +1,500 @@
+     * {@inheritdoc}
+     * 
+     */
+    public function beginTransaction()
```
Can you explain all this transaction code for?
At first glance it seems to support nested transactions, but  i fail see how to reach a transaction depth > 1.

kimhemsoe commented on lib/Doctrine/DBAL/Driver/Ibase/AbstractIbaseConnection.php:
```diff
@@ -0,0 +1,500 @@
+     * @param type $attribute
+     * @return type
+     */
+    public function getAttribute($attribute)
```
Not used anywhere

kimhemsoe commented on lib/Doctrine/DBAL/Driver/Ibase/AbstractIbaseConnection.php:
```diff
@@ -0,0 +1,500 @@
+                    $this->attrDcTransWait = $value;
+                    break;
+                }
+            case \PDO::ATTR_AUTOCOMMIT: {
```
What happens if you change to auto commit inside a transaction?

kimhemsoe commented on lib/Doctrine/DBAL/Driver/Ibase/AbstractIbaseConnection.php:
```diff
@@ -0,0 +1,500 @@
+     * @param type $params
+     * @return type
+     */
+    protected function makeDbString($params)
```
private

kimhemsoe commented on lib/Doctrine/DBAL/Driver/Ibase/AbstractIbaseConnection.php:
```diff
@@ -0,0 +1,500 @@
+     */
+    public function getExecuteMode()
+    {
+        return $this->executeMode;
```
not defined

kimhemsoe commented on lib/Doctrine/DBAL/Driver/Ibase/AbstractIbaseConnection.php:
```diff
@@ -0,0 +1,500 @@
+     *
+     * @return integer
+     */
+    public function getExecuteMode()
```
not used

kimhemsoe commented on lib/Doctrine/DBAL/Driver/Ibase/AbstractIbaseConnection.php:
```diff
@@ -0,0 +1,500 @@
+        $result = "";
+        switch ($isolationLevel) {
+            case \Doctrine\DBAL\Connection::TRANSACTION_READ_UNCOMMITTED: {
+                    $result .= 'SET TRANSACTION READ WRITE ISOLATION LEVEL READ COMMITTED RECORD_VERSION';
```
Why concatenation?

kimhemsoe commented on lib/Doctrine/DBAL/Driver/Ibase/AbstractIbaseConnection.php:
```diff
@@ -0,0 +1,500 @@
+            case \Doctrine\DBAL\Connection::TRANSACTION_SERIALIZABLE: {
+                    $result .= 'SET TRANSACTION READ WRITE ISOLATION LEVEL SNAPSHOT TABLE STABILITY';
+                    break;
+                }
```
Missing default case

kimhemsoe commented on lib/Doctrine/DBAL/Driver/Ibase/AbstractIbaseStatement.php:
```diff
@@ -0,0 +1,657 @@
+ * 
+ * @author Andreas Prucha, Helicon Software Development <prucha@helicon.co.at>
+ */
+abstract class AbstractIbaseStatement implements \IteratorAggregate, Statement
```
Please change attributes to private where possible.

kimhemsoe replied:
And methods to private where possible.

kimhemsoe commented on lib/Doctrine/DBAL/Driver/Ibase/AbstractIbaseStatement.php:
```diff
@@ -0,0 +1,657 @@
+     * 
+     * @param string $statement
+     */
+    protected function setStatement($statement)
```
private

kimhemsoe commented on lib/Doctrine/DBAL/Driver/Ibase/AbstractIbaseStatement.php:
```diff
@@ -0,0 +1,657 @@
+     * @param type $aPropertyList Associative array of object properties
+     * @return object created object or object passed in $aClassOrObject
+     */
+    protected function createObjectAndSetPropertiesCaseInsenstive($aClassOrObject, array $aConstructorArgArray, array $aPropertyList)
```
private

kimhemsoe commented on lib/Doctrine/DBAL/Platforms/AbstractFbIbPlatform.php:
```diff
@@ -0,0 +1,1378 @@
+ * 
+ * @author Andreas Prucha, Helicon Software Development <prucha@helicon.co.at>
+ */
+abstract class AbstractFbIbPlatform extends AbstractPlatform
```
Please change protected methods to private where possible.

kimhemsoe commented on lib/Doctrine/DBAL/Driver/Ibase/AbstractIbaseStatement.php:
```diff
@@ -0,0 +1,657 @@
+            $this->queryParamBindings[$column - 1] = &$variable;
+            $this->queryParamTypes[$column - 1] = $type;
+        } else {
+            if (isset($this->namedParamsMap[$column])) {
```
Named parameters is not a requirement. Dbal\Connection will never use it. You can only use it directly from the driver connection directly. I suggest you remove it, makes maintenance easier further down the road.

kimhemsoe commented on tests/Doctrine/Tests/DBAL/Functional/TableGeneratorTest.php:
```diff
@@ -14,6 +14,8 @@ class TableGeneratorTest extends \Doctrine\Tests\DbalFunctionalTestCase
     {
         parent::setUp();
+        
+//        $this->skipOnDriverOrPlatform('firebird', 'TableGenerator does not work - Needs to be figured out why', true);
```
Remove?

marc-x commented:
+1

mathielen commented:
@ancpru: any change you can comment on the review, so we might have a change to get this merged?

dimitrov-adrian commented:
+1

JuergenFB commented:
Did anybody still support this issue?

I looks like nothing has done since October 2015 ...

barricade823 commented:
+1

bambucha4 commented:
Is it possible to introduce this functionality in the near future?

Majkl578 commented:
I think so, yes, someone just needs to finish the PR into a mergeable state.

Right now, this includes adressing comments from the review here, rebasing on top of current master (2.9-dev), adding it to our CI matrix (at least Travis) and making sure it works with Firebird 3 + 4.

Bonus point re-test with PDO, whether it's usable or not (the issue referenced in OP is 3 years old):
> The driver uses the ibase-api. Originally I also had a PDO-based implementation. Despite it worked quite well in a real world application, it terribly failed in the Doctrine test suite because of strange transaction behaviour of the Firebird-PDO, thus I finally removed it.

If anyone wants to work on this, feel free to create new PR, re-using commits from here. 👍 

Note that we don't need any focus on InterBase support as it's a proprietary closed-source DB, only Firebird.

Ocramius commented:
> I think so, yes, someone just needs to finish the PR into a mergeable state.

We'd also need someone experienced with Firebird to join the DBAL team for maintenance purposes: Firebird is very exotic technology (to put it in a nice way), so we won't be able to help with bugs or generic issues if we're not having the expertise on board.
